src/hotspot/share/gc/shared/gcArguments.cpp

```diff
@@ -138,6 +141,15 @@ void GCArguments::initialize_heap_flags_and_sizes() {
     vm_exit_during_initialization("Too small minimum heap");
   }
 
+#if defined(LINUX)
+  if (OSContainer::is_containerized() && FLAG_IS_CMDLINE(InitialHeapSize)) {
+    jlong limit = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((limit > 0) && (InitialHeapSize >= (julong) limit)) {
```

Comment:
Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.

Reply:
> Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.

What will happen with `-Xms7G` on a machine configured with 4G physical memory + 4G swap space?
Does `os::physical_memory()`  return `physical memory + swap space size`?

Reply:
@jerboaa Good point. Also take a look at Arguments::set_heap_size() and how that function uses os::physical_memory() and also limit_heap_by_allocatable_memory(). It seems like any change in GCArguments should go hand-in-hand with the code in set_heap_size.

Reply:
> > Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.
> 
> What will happen with `-Xms7G` on a machine configured with 4G physical memory + 4G swap space? Does `os::physical_memory()` return `physical memory + swap space size`?

It doesn't. What would happen on a physical machine, though? It should really do the same thing regardless of the container settings. Container settings, in a nutshell, should be an abstraction of "physical" host. Otherwise behave the same. Having said that, in containers (at least kubernetes) swap is usually disabled. Also, does it make sense to allow `-Xms7g` on a system with `4G` of physical RAM? I'm doubtful.

Reply:
> @jerboaa Good point. Also take a look at Arguments::set_heap_size() and how that function uses os::physical_memory() and also limit_heap_by_allocatable_memory(). It seems like any change in GCArguments should go hand-in-hand with the code in set_heap_size.

+1

Reply:
I agree with the comments that this should behave exactly as it would for a physical machine. The fact we are in a container is not really relevant.

Reply:
Maybe we ought to move the check into arguments.cpp, and compare os::physical_memory against (initial heap size + initial code cache size + .... something else)?

Forcing the VM to quit may have backward compatibility issues, though. Some apps may set their -Xms to be larger than the container memory limit, but they may use so little heap memory that the VM never commits enough memory to trigger the OOM killer. With this PR those apps will suddenly start failing.